capsules/src/pwm.rs

```diff
@@ -0,0 +1,150 @@
+
+// An empty app, for potential uses in future updates of the driver
+#[derive(Default)]
+pub struct App;
+
+pub struct Pwm<'a, const NUM_PINS: usize> {
+    /// The usable pwm pins.
+    pwm_pins: &'a [&'a dyn hil::pwm::PwmPin; NUM_PINS],
+    /// Per-app state.
+    apps: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    /// An array of apps associated to their reserved pins.
+    active_process: [OptionalCell<ProcessId>; NUM_PINS],
+}
+
+impl<'a, const NUM_PINS: usize> Pwm<'a, NUM_PINS> {
+    pub fn new(
+        pwm_pins: &'a [&'a dyn hil::pwm::PwmPin; NUM_PINS],
+        grant: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<0>>,
+    ) -> Pwm<'a, NUM_PINS> {
+        assert!(NUM_PINS <= (u16::MAX as usize));
+        const EMPTY: OptionalCell<ProcessId> = OptionalCell::empty();
+        Pwm {
+            pwm_pins: pwm_pins,
+            apps: grant,
+            active_process: [EMPTY; NUM_PINS],
+        }
+    }
+
+    pub fn claim_pin(&self, processid: ProcessId, pin: usize) -> bool {
+        // Attempt to get the app that is using the pin.
+        self.active_process[pin].map_or(true, |id| {
+            // If the app is empty, that means that there is no app currently using this pin,
+            // therefore the pin could be usable by the new app
+            if id == &processid {
+                // The same app is trying to access the pin it has access to, valid
+                true
+            } else {
+                // An app is trying to access another app's pin, invalid
+                false
+            }
+        })
+    }
+
+    pub fn release_pin(&self, pin: usize) {
+        // Release the claimed pin so that it can now be used by another process.
+        self.active_process[pin].clear();
+    }
+}
+
+/// Provide an interface for userland.
+impl<'a, const NUM_PINS: usize> SyscallDriver for Pwm<'a, NUM_PINS> {
+    /// Command interface.
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Return number of PWM pins if this driver is included on the platform.
+    /// - `1`: Start the PWM pin output. First 16 bits of `data1` are used for the duty cycle, as a
+    ///     percentage with 2 decimals, and the last 16 bits of `data1` are used for the PWM channel
+    ///     to be controlled. `data2` is used for the frequency in hertz. For the duty cycle, 100% is
+    ///     the max duty cycle for this pin.
+    /// - `2`: Stop the PWM output.
+    /// - `3`: Return the maximum possible frequency for this pin.
+    fn command(
+        &self,
+        command_num: usize,
+        data1: usize,
+        data2: usize,
+        processid: ProcessId,
+    ) -> CommandReturn {
+        match command_num {
+            // Check whether the driver exists.
+            0 => CommandReturn::success(),
+
+            // Return number of usable PWM pins.
+            1 => CommandReturn::success_u32(NUM_PINS as u32),
+
+            // Start the pwm output.
+
+            // data1 format
```

Comment:
```suggestion
            // data1 stores the duty cycle and the pin number in the format
            // data2 stores the frequency
```

Please add the explanation about why you chose this format